gpio: gd32: initial support

gmarull opened the pull request:
This PR introduces a GPIO driver for GigaDevice SoCs, including AF (e.g. F4XX) and AFIO (e.g. F403) variants.

Tested on:

- GD32F450I-EVAL
  - samples/basic/blinky
  - samples/basic/button
  - tests/drivers/gpio/gpio_basic_api
  - tests/drivers/gpio/gpio_api_1pin

- GD32VF103V-EVAL (not mainlined yet, https://github.com/teslabs/zephyr/tree/gpio-gd32-vf103)
  - samples/basic/blinky
  - samples/basic/button
  - tests/drivers/gpio/gpio_basic_api
  - tests/drivers/gpio/gpio_api_1pin
 
- GD32F403Z-EVAL (@nandojve)
  - samples/basic/blinky
  - samples/basic/button

gmarull committed:
dts: arm: gigadevice: gd32f4xx: fix gpio support

- The gpio dt-bindings were not included
- gpio-controller property was not present in gpio nodes
- number of gpio cells was not present in gpio nodes

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: arm: gigadevice: gd32f403: fix gpio support

- The gpio dt-bindings were not included
- gpio-controller property was not present in gpio nodes
- number of gpio cells was not present in gpio nodes

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
drivers: gpio: gd32: initial version

Add GPIO driver for Gigadevice SoCs. The driver supports devices with
the AF and AFIO models.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: gd32f450i_eval: add leds and keys

Add definitions for board LEDs and keys now that a GPIO driver exists.
The GPIO driver has also been enabled in the board defconfig, since it
is required to run some basic samples (e.g. blinky).

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: gd32f403z_eval: add leds and keys

Add definitions for board LEDs and keys now that a GPIO driver exists.
The GPIO driver has also been enabled in the board defconfig, since it
is required to run some basic samples (e.g. blinky).

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull commented:
@cameled FYI

nandojve commented on drivers/gpio/gpio_gd32.c:
```diff
@@ -0,0 +1,221 @@
+{
+	const struct gpio_gd32_config *config = port->config;
+
+#ifdef GD32_HAS_AF_PINMUX
```
I think it will be better if we adopt IS_ENABLED when possible.

```
if (IS_ENABLED(GD32_HAS_AF_PINMUX)) {
...
} else {
...
}
```

gmarull replied:
This strategy can't be used here. This implies code from both branches can be always compiled, not possible here due to non-shared registers (i.e. definitions).

nandojve commented on drivers/gpio/gpio_gd32.c:
```diff
@@ -0,0 +1,221 @@
+{
+	const struct gpio_gd32_config *config = port->config;
+
+	GPIO_OCTL(config->reg) = (GPIO_OCTL(config->reg) & ~mask) |
+				 (value & mask);
```
```
	GPIO_OCTL(config->reg) = (GPIO_OCTL(config->reg) & ~mask)
			       | (value & mask);
```

gmarull replied:
adjusted formatting

nandojve commented on drivers/gpio/gpio_gd32.c:
```diff
@@ -0,0 +1,221 @@
+{
+	const struct gpio_gd32_config *config = port->config;
+
+#ifdef GD32_HAS_AF_PINMUX
```
`if (IS_ENABLED(GD32_HAS_AF_PINMUX)) {`

gmarull replied:
ditto

cameled commented:
Hi @gmarull, can we add pin default value setting for output mode? like: https://github.com/zephyrproject-rtos/zephyr/blob/38c42f69d28b8ce7cefb2b1a265400e83dd654aa/drivers/gpio/gpio_nrfx.c#L214 

spi cs_control need this feature to disable the slave device by default.

gmarull committed:
drivers: interrupt_controller: initial support for GD32 EXTI

Add initial support for the GigaDevice External Interrupt Controller.
This driver is required to manage GPIO interrupts. Only EXTI lines 0 to
15 are supported for now (no LVD, RTC, etc.). Driver can be extended in
the future to add support for extra EXTI lines.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

cameled commented on drivers/gpio/gpio_gd32.c:
```diff
@@ -0,0 +1,221 @@
+
+#define DT_DRV_COMPAT gd_gd32_gpio
+
+#include <drivers/gpio.h>
```
Please add soc.h in here, gd32f3x0 need it to explicit include the gd32f3x0_gpio.h.

gmarull replied:
done

gmarull committed:
dts: arm: gigadevice: gd32f403: add support for exti

Add DT node for EXTI (External Interrupt Controller). Only line
interrupts added for now (used for GPIO interrupts).

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: arm: gigadevice: gd32f4xx: add support for exti

Add DT node for EXTI (External Interrupt Controller). Only line
interrupts added for now (used for GPIO interrupts).

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: bindings: interrupt-controller: add gd32 exti

Add binding for the GigaDevice EXTI interrupt controller.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
tests: drivers: gpio: gpio_basic_api: add support for gd32f450i_eval

Add DT overlay to run gpio_basic_api test on GD32F450I-EVAL board. In
order to make testing easy a couple of accessible pins have been
selected: PC6 and PC7, the last 2 pins on the DCI camera 2.54" header.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull commented:
> Hi @gmarull, can we add pin default value setting for output mode? like:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/38c42f69d28b8ce7cefb2b1a265400e83dd654aa/drivers/gpio/gpio_nrfx.c#L214
> 
> spi cs_control need this feature to disable the slave device by default.

done

gmarull commented:
@nandojve would appreciate testing on F403, since it uses the AFIO model (implementation not validated on my side)

cameled commented on drivers/gpio/gpio_gd32.c:
```diff
@@ -0,0 +1,363 @@
+
+#ifdef CONFIG_GD32_HAS_AF_PINMUX
+	/* enable access to SYSCFG_EXTISSn registers */
+	rcu_periph_clock_enable(RCU_SYSCFG);
```
For AF_PINMUX, RCU_SYSCFG are not always exist, GD32F3x0 doesn't have this clock control bit. 
Not sure which way is more general, just leave a comment in here.

gmarull replied:
solved by using DT (check syscfg node)

gmarull committed:
dts: bindings: misc: add gd32 syscfg

Add support for GigaDevice SYSCFG.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: arm: gigadevice: gd32f4xx: add syscfg node

Add node for the SYSCFG registers (required by the GPIO driver).

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull commented on drivers/interrupt_controller/intc_gd32_exti.c:
```diff
@@ -0,0 +1,194 @@
+/*
```
Note: I'm not sure if implementing EXTI as a sort of "interrupt controller" is in the end a good idea. EXTI isn't really an interrupt controller on its own (NVIC is used), it's just a set of registers that allow to configure multiple external interrupt sources, being GPIO one of them. I'll try embedding this directly into GPIO and compare the result.

nandojve replied:
SAM0 uses NVIC line too. What is important know is EXTI depends on GPIO ? I mean, EXTI only works from lines [0-15] when GPIO clock is enabled?

gmarull replied:
Dropping the idea for now, I think it's acceptable as it is now.

nandojve commented on boards/arm/gd32f403z_eval/doc/index.rst:
```diff
@@ -58,6 +58,8 @@ The board configuration supports the following hardware features:
 +-----------+------------+-----------------------+
 | Interface | Controller | Driver/Component      |
 +===========+============+=======================+
+| GPIO      | on-chip    | gpio                  |
++-----------+------------+-----------------------+
```
Should EXTI be added ?

gmarull replied:
will do

gmarull replied:
done

nandojve commented on boards/arm/gd32f450i_eval/doc/index.rst:
```diff
@@ -59,6 +59,9 @@ The board configuration supports the following hardware features:
    * - Peripheral
      - Kconfig option
      - Devicetree compatible
+   * - GPIO
+     - :kconfig:`CONFIG_GPIO`
+     - :dtcompatible:`gd,gd32-gpio`
```
same...

nandojve reviewed (changes requested):
Hi @gmarull ,
Looks good!

I tested with success samples/basic/[led/button].
Just few comment and I was wondering if it is possible add `gpio_basic_api` sample to F403 to test both AF/AFIO.

gmarull commented on boards/arm/gd32f450i_eval/gd32f450i_eval_defconfig:
```diff
@@ -12,3 +12,5 @@ CONFIG_CORTEX_M_SYSTICK=y
 CONFIG_CONSOLE=y
 CONFIG_UART_CONSOLE=y
 CONFIG_SERIAL=y
+
+CONFIG_GPIO=y
```
As far as I know, UART and GPIO and considered part of a minimum set (GPIO allows running the blinky sample, part of the getting started). Of course, none of both are minimal: many applications don't use UART in production either.

gmarull committed:
tests: drivers: gpio: gpio_basic_api: add support for gd32f403z_eval

Add DT overlay to run gpio_basic_api test on GD32F403Z-EVAL board. In
order to make testing easy a couple of accessible pins have been
selected: PD0 and PD1 pins exposed via P3 and P2 respectively.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull commented:
> Hi @gmarull , Looks good!
> 
> I tested with success samples/basic/[led/button]. Just few comment and I was wondering if it is possible add `gpio_basic_api` sample to F403 to test both AF/AFIO.

Added overlay for F403Z-EVAL. You should be able to test it by bridging the 2 DAC pins with a jumper (see overlay details).

nandojve commented:
Hi @gmarull ,

> Added overlay for F403Z-EVAL. You should be able to test it by bridging the 2 DAC pins with a jumper (see overlay details).

Not sure why, GPIO at DAC port fail.
My config is: JP12 1-2 and JP7 1-2

I switch to port D 0-1 P2 + P3 which result is:

```
Running test suite gpio_basic_test
===================================================================
START - test_gpio_port
Validate device GPIOD
Check GPIOD output 0 connected to input 1

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/gpio/gpio_basic_api/src/test_gpio_port.c:99: setup: (v1 & BIT(PIN_IN) not equal to 0)
out low does not read low
 FAIL - test_gpio_port in 0.21 seconds
===================================================================
START - test_gpio_callback_add_remove
callback_2 triggered: 1
callback_1 triggered: 1
callback_2 triggered: 1
callback_2 triggered: 2
FAIL - test_callback_add_remove@113. not trigger callback correctly

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/gpio/gpio_basic_api/src/test_callback_manage.c:235: test_gpio_callback_add_remove: test_callback_add_remove() not equal to TC_PASS

 FAIL - test_gpio_callback_add_remove in 2.432 seconds
===================================================================
START - test_gpio_callback_self_remove
callback_remove_self triggered: 1
callback_1 triggered: 1
callback_1 triggered: 1
callback_1 triggered: 2
FAIL - test_callback_self_remove@174. not trigger remaining callback correctly

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/gpio/gpio_basic_api/src/test_callback_manage.c:241: test_gpio_callback_self_remove: test_callback_self_remove() not equal to TC_PASS

 FAIL - test_gpio_callback_self_remove in 2.534 seconds
===================================================================
START - test_gpio_callback_enable_disable
callback_2 triggered: 1
callback_1 triggered: 1
callback_2 triggered: 1
callback_1 triggered: 1
 PASS - test_gpio_callback_enable_disable in 3.610 seconds
===================================================================
START - test_gpio_callback_variants
callback triggered: 1
OUT init a01, IN cfg 34000, cnt 1
callback triggered: 1
OUT init 600, IN cfg 54000, cnt 1
callback triggered: 1
OUT init 600, IN cfg 5c000, cnt 1
callback triggered: 1
OUT init a01, IN cfg 3c000, cnt 1
Mode 44000 not supported
Mode 24000 not supported
Mode 4c000 not supported
Mode 2c000 not supported
callback triggered: 1
callback triggered: 2
OUT init a01, IN cfg 74000, cnt 2
 PASS - test_gpio_callback_variants in 5.537 seconds
===================================================================
Test suite gpio_basic_test failed.
===================================================================
PROJECT EXECUTION FAILED
```

gmarull commented:
> Hi @gmarull ,
> 
> > Added overlay for F403Z-EVAL. You should be able to test it by bridging the 2 DAC pins with a jumper (see overlay details).
> 
> Not sure why, GPIO at DAC port fail. My config is: JP12 1-2 and JP7 1-2
> 
> I switch to port D 0-1 P2 + P3 which result is:
> 

Are P2/P3 connected to 1-2 (EXMC signals) and display removed from the board?

![image](https://user-images.githubusercontent.com/25011557/145721473-8952278b-f603-404a-9f03-0926056b6705.png)

nandojve commented:
> Are P2/P3 connected to 1-2 (EXMC signals) and display removed from the board (to bridge EXMC_D2/EXMC_D3) ?

PD1-PD0 are connected by P2-2 + P3-2 using a jumper. No need to remove LCD.

cameled commented on drivers/gpio/gpio_gd32.c:
```diff
@@ -0,0 +1,369 @@
+	static struct gpio_gd32_data gpio_gd32_data##n;			       \
+									       \
+	DEVICE_DT_INST_DEFINE(n, &gpio_gd32_init, NULL, &gpio_gd32_data##n,    \
+			      &gpio_gd32_config##n, PRE_KERNEL_1,	       \
```
GPIO driver initialize at PRE_KERNEL_1 level. Others driver, like spi and adc, normally initialize at POST_KERNEL level.
when they call pinctrl_apply_state function to adjust pin function, there may have a risk to incorrectly disable the GPIO.

https://github.com/zephyrproject-rtos/zephyr/blob/521e6ffb7c1fcb663b98546eeba7822e03ca5a49/drivers/pinctrl/pinctrl_gd32_af.c#L98

gmarull replied:
Fixed, thanks for spotting. We can likely improve this once a proper clock control API is in place.

gmarull committed:
drivers: pinctrl: gd32: do not disable clock

Disabling clock may conflict with concurrent usage of GPIOs and pins in
AF mode. This can be improved once a proper clock control API is in
place.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull commented:
> Hi @gmarull ,
> 
> > Added overlay for F403Z-EVAL. You should be able to test it by bridging the 2 DAC pins with a jumper (see overlay details).
> 
> Not sure why, GPIO at DAC port fail. My config is: JP12 1-2 and JP7 1-2
> 
> I switch to port D 0-1 P2 + P3 which result is:
> 
> ```
> Running test suite gpio_basic_test
> ===================================================================
> START - test_gpio_port
> Validate device GPIOD
> Check GPIOD output 0 connected to input 1
> 
>     Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/gpio/gpio_basic_api/src/test_gpio_port.c:99: setup: (v1 & BIT(PIN_IN) not equal to 0)
> out low does not read low
>  FAIL - test_gpio_port in 0.21 seconds
> ===================================================================
> START - test_gpio_callback_add_remove
> callback_2 triggered: 1
> callback_1 triggered: 1
> callback_2 triggered: 1
> callback_2 triggered: 2
> FAIL - test_callback_add_remove@113. not trigger callback correctly
> 
>     Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/gpio/gpio_basic_api/src/test_callback_manage.c:235: test_gpio_callback_add_remove: test_callback_add_remove() not equal to TC_PASS
> 
>  FAIL - test_gpio_callback_add_remove in 2.432 seconds
> ===================================================================
> START - test_gpio_callback_self_remove
> callback_remove_self triggered: 1
> callback_1 triggered: 1
> callback_1 triggered: 1
> callback_1 triggered: 2
> FAIL - test_callback_self_remove@174. not trigger remaining callback correctly
> 
>     Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/gpio/gpio_basic_api/src/test_callback_manage.c:241: test_gpio_callback_self_remove: test_callback_self_remove() not equal to TC_PASS
> 
>  FAIL - test_gpio_callback_self_remove in 2.534 seconds
> ===================================================================
> START - test_gpio_callback_enable_disable
> callback_2 triggered: 1
> callback_1 triggered: 1
> callback_2 triggered: 1
> callback_1 triggered: 1
>  PASS - test_gpio_callback_enable_disable in 3.610 seconds
> ===================================================================
> START - test_gpio_callback_variants
> callback triggered: 1
> OUT init a01, IN cfg 34000, cnt 1
> callback triggered: 1
> OUT init 600, IN cfg 54000, cnt 1
> callback triggered: 1
> OUT init 600, IN cfg 5c000, cnt 1
> callback triggered: 1
> OUT init a01, IN cfg 3c000, cnt 1
> Mode 44000 not supported
> Mode 24000 not supported
> Mode 4c000 not supported
> Mode 2c000 not supported
> callback triggered: 1
> callback triggered: 2
> OUT init a01, IN cfg 74000, cnt 2
>  PASS - test_gpio_callback_variants in 5.537 seconds
> ===================================================================
> Test suite gpio_basic_test failed.
> ===================================================================
> PROJECT EXECUTION FAILED
> ```

Should be fixed now (tested on GD32VF103V-EVAL which has the same GPIO IP).

mnkp reviewed (approved):
Comment to "drivers: gpio: gd32: initial version" mentions "TODO: Add support for interrupts.". However support for interrupts is already implemented. Could be corrected.

gmarull commented:
> Comment to "drivers: gpio: gd32: initial version" mentions "TODO: Add support for interrupts.". However support for interrupts is already implemented. Could be corrected.

removed TODO, thanks for spotting

nandojve reviewed (approved):
GD32F403Z-EVAL all green
  - samples/basic/blinky
  - samples/basic/button
  - tests/drivers/gpio/gpio_basic_api
  - tests/drivers/gpio/gpio_api_1pin

cfriedt commented:
@gmarull - can you please rebase and fix merge conflicts?

gmarull commented:
> @gmarull - can you please rebase and fix merge conflicts?

done
